stateproof/worker_test.go

```diff
@@ -546,3 +550,104 @@ func TestSignerDoesntDeleteKeysWhenDBDoesntStoreSigs(t *testing.T) {
 }
+
+// TestSigBroacastTwoPerSig checks if each signature is broadcasted twice per period
+// It generates numAddresses and prepares a database with the account/signatures.
+// Then, calls broadcastSigs with round numbers spanning periods and
+// makes sure each account has 2 sigs sent per period if originated locally, and 1 sig
+// if received from another relay.
+func TestSigBroacastTwoPerSig(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	// Prepare the addresses and the keys
+	numAddresses := 10
+	signatureBcasted := make(map[basics.Address]int)
+	fromThisNode := make(map[basics.Address]bool)
+	var keys []account.Participation
+	for i := 0; i < numAddresses; i++ {
+		var parent basics.Address
+		crypto.RandBytes(parent[:])
+		p := newPartKey(t, parent)
+		defer p.Close()
+		keys = append(keys, p.Participation)
+		signatureBcasted[parent] = 0
+	}
+
+	s := newWorkerStubs(t, keys, len(keys))
+	spw := newTestWorker(t, s)
+	proto := config.Consensus[protocol.ConsensusFuture]
+
+	// Prepare the database
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		return initDB(tx)
+	})
+	require.NoError(t, err)
+
+	// All the keys are for round 255. This way, starting the period at 256,
+	// there will be no disqualified signatures from broadcasting because they are
+	// into the future.
+	round := basics.Round(255)
+
+	// Sign the message
+	spRecords := stateProofRecordFromKeys(round, keys)
+	sigs := make([]sigFromAddr, 0, len(keys))
+	stateproofMessage := stateproofmsg.Message{}
+	hashedStateproofMessage := stateproofMessage.IntoStateProofMessageHash()
+	for _, key := range spRecords {
+		sig, err := key.StateProofSecrets.SignBytes(hashedStateproofMessage[:])
+		require.NoError(t, err)
+		sigs = append(sigs, sigFromAddr{
+			SignerAddress: key.Account,
+			Round:         round,
+			Sig:           sig,
+		})
+	}
+
+	// Add the signatures to the databse
+	ftn := true
+	for _, sfa := range sigs {
+		err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+			return addPendingSig(tx, sfa.Round, pendingSig{
+				signer:       sfa.SignerAddress,
+				sig:          sfa.Sig,
+				fromThisNode: ftn,
+			})
+		})
+		require.NoError(t, err)
+		fromThisNode[sfa.SignerAddress] = ftn
+		// alternate the from this node argument between addresses
+		ftn = !ftn
+	}
+
+	// Collect the broadcast messages
+	tns := spw.net.(*testWorkerStubs)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for bMsg := range tns.sigmsg {
+			sfa := sigFromAddr{}
+			err = protocol.Decode(bMsg, &sfa)
+			require.NoError(t, err)
+			signatureBcasted[sfa.SignerAddress]++
+		}
+	}()
+
+	// Broadcast the messages
+	periods := 5
+	for brnd := 257; brnd < 257+256*periods; brnd++ {
+		spw.broadcastSigs(basics.Round(brnd), proto)
+	}
+
+	close(tns.sigmsg)
+	wg.Wait()
+
+	// Verify the number of times each signature was broadcast
+	for addr, sb := range signatureBcasted {
+		if fromThisNode[addr] {
+			require.Equal(t, 2*periods, sb)
```

Comment:
I'm not certain this test makes certain each interval has exactly two broadcasts per address that are from the node.
can you change it to be a more granular test? That is, check that per period 2 signatures are present, and not an aggregation of all period results?